addons/binding/org.openhab.binding.s7/src/main/java/org/openhab/binding/s7/handler/S7BridgeHandler.java

```diff
@@ -0,0 +1,374 @@
+
+        byte[] dataPA = new byte[PALength];
+        byte[] dataPE = new byte[PELength];
+        byte[] dataDB = new byte[DBLength];
+        byte[] dataMK = new byte[MKLength];
+
+        if (client.ReadArea(S7.S7AreaPA, 1, 0, PALength, dataPA) == 0) {
+            data.put(S7.S7AreaPA, dataPA);
+            nReadError = 0;
+        } else {
+            logger.error(S7Client.ErrorText(client.LastError));
+            nReadError++;
+        }
+
+        if (client.ReadArea(S7.S7AreaPE, 1, 0, PELength, dataPE) == 0) {
+            data.put(S7.S7AreaPE, dataPE);
+            nReadError = 0;
+        } else {
+            logger.error(S7Client.ErrorText(client.LastError));
+            nReadError++;
+        }
+
+        if (client.ReadArea(S7.S7AreaDB, 1, 0, DBLength, dataDB) == 0) {
+            data.put(S7.S7AreaDB, dataDB);
+            nReadError = 0;
+        } else {
+            logger.error(S7Client.ErrorText(client.LastError));
+            nReadError++;
+        }
+
+        if (client.ReadArea(S7.S7AreaMK, 1, 0, MKLength, dataMK) == 0) {
+            data.put(S7.S7AreaMK, dataMK);
+            nReadError = 0;
+        } else {
+            logger.error(S7Client.ErrorText(client.LastError));
+            nReadError++;
+        }
+
+        if (nReadError == 20) {
+            this.destroyClient();
+            this.createClient();
+        }
+
+        return data;
+    }
+
+    public void setBit(int Area, int Address, boolean value) {
+        byte Data[] = new byte[1];
+
+        synchronized (client) {
+            client.ReadArea(Area, 1, Address / 8, 1, Data);
+            S7.SetBitAt(Data, 0, Address % 8, value);
+            client.WriteArea(Area, 1, Address / 8, 1, Data);
+        }
+    }
+
+    private boolean initializingState = true;
+    private boolean runningState = false;
+
+    private void onConnectionEstablished() {
+        if (!runningState || initializingState) {
+            runningState = true;
+            logger.info("Connection to S7 PLC {} established.", getConfig().get(HOST));
+
+            updateStatus(ThingStatus.ONLINE);
+
+            logger.info("S7 bridge {} now online.", getConfig().get(HOST));
+        }
+    }
+
+    private void onConnectionLost() {
+        if (runningState || initializingState) {
+            runningState = false;
+            logger.info("Connection to S7 PLC {} lost. Trying to reconnect in 5 seconds.", getConfig().get(HOST));
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.OFFLINE.COMMUNICATION_ERROR,
+                    S7Client.ErrorText(client.LastError));
+
+            new Thread() {
+                @Override
+                public void run() {
+                    while (!client.Connected) {
+                        try {
+                            Thread.sleep(5000);
+                        } catch (InterruptedException e) {
+                            e.printStackTrace();
```

Comment:
If we receive this InterruptedException, the interrupted flag was cleared and we most likely need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/